core/src/main/scala/org/broadinstitute/dsde/rawls/fastpass/FastPassService.scala

```diff
@@ -0,0 +1,444 @@
+      fastPassGrants.groupBy(g => (g.accountEmail, g.accountType, g.resourceType, g.resourceName))
+    val iamUpdates = grantsByUserAndResource.toSeq.map { grouped =>
+      val ((accountEmail, accountType, resourceType, resourceName), grants) = grouped
+      val organizationRoles = grants.map(_.organizationRole).toSet
+      val workbenchEmail = WorkbenchEmail(accountEmail.value)
+      val googleProject = GoogleProject(googleProjectId.value)
+      () =>
+        removeFastPassGrants(resourceType, resourceName, workbenchEmail, accountType, organizationRoles, googleProject)
+    }
+    for {
+      // this `flatMap` is necessary to run the IAM Updates in sequence, as to not sent conflicting policy updates to GCP
+      _ <- DBIO.from(iamUpdates.foldLeft(Future.successful())((a, b) => a.flatMap(_ => b())))
+      _ <- DBIO.seq(fastPassGrants.map(_.id).map(removeGrantFromDb): _*)
+    } yield ()
+  }
+
+  private def removeFastPassGrants(resourceType: IamResourceType,
+                                   resourceName: String,
+                                   workbenchEmail: WorkbenchEmail,
+                                   memberType: IamMemberType,
+                                   organizationRoles: Set[String],
+                                   googleProject: GoogleProject
+  ): Future[Unit] = {
+    logger.info(
+      s"Removing FastPass IAM bindings for ${workbenchEmail.value} on ${resourceType.value} $resourceName"
+    )
+    for {
+      _ <- resourceType match {
+        case IamResourceTypes.Bucket =>
+          googleStorageDAO.removeIamRoles(
+            GcsBucketName(resourceName),
+            workbenchEmail,
+            memberType,
+            organizationRoles,
+            userProject = Some(googleProject)
+          )
+        case IamResourceTypes.Project =>
+          googleIamDao.removeRoles(
+            googleProject,
+            workbenchEmail,
+            memberType,
+            organizationRoles
+          )
+      }
+      _ <- openTelemetry
+        .incrementCounter(s"fastpass-iam-revoked-${memberType.value}", tags = openTelemetryTags)
+        .unsafeToFuture()
+    } yield ()
+  }
+
+  private def setupProjectRoles(workspace: Workspace,
+                                samResourceRoles: Set[SamResourceRole],
+                                userAndPet: UserAndPetEmails,
+                                expirationDate: DateTime
+  ): ReadWriteAction[Unit] = {
+    val projectIamRoles = samResourceRoles.flatMap(samWorkspaceRoleToGoogleProjectIamRoles)
+    val condition = conditionFromExpirationDate(expirationDate)
+
+    for {
+      _ <- DBIO.from(addUserAndPetToProjectIamRoles(workspace.googleProjectId, projectIamRoles, userAndPet, condition))
+      _ <- writeGrantsToDb(
+        workspace.workspaceId,
+        userAndPet,
+        gcpResourceType = IamResourceTypes.Project,
+        workspace.googleProjectId.value,
+        projectIamRoles,
+        expirationDate
+      )
```

Comment:
example of where mixing `DBIO` and `Future` is a little touchy. The roles have been added to google yet future errors may trigger a rollback and the records will not be written to the db.

Reply:
what I am getting at here is that the order this code is written is not necessarily the order it is executed. I think no matter the order these commands are written here, google will be updated before the transaction commits and even if the transaction rolls back.